### PR TITLE
feat(cloud): Introduce connections to proxy unikernel

### LINKED ISSUE #1495: Ability to "tunnel" from localhost to service group or port forward to an instance
### Feature request summary

This issue tracks the ability to let the user access the internal network and/or internally exposed ports of application services within a KraftCloud service group.

The purpose of this is to let users a). connect to services which are typically not exposing TLS-based traffic, which KraftCloud only supports and b). debug instances without having to publicly expose the port.

This should manifest itself as a new CLI subcommand:

```
kraft cloud tunnel <SERVICE|INSTANCE> [PORT[:PORT[/PROTOCOL]
```

Where then:

- `kraft cloud tunnel my-service` -- forwards everything, you can do `ping` on instances in the same service group. Needs some thinking and sounds more complex to do.
- `kraft cloud tunnel my-instance 8080` -- forwards port 8080 on localhost to 8080 of the instance `my-instance`.
- `kraft cloud tunnel my-instance 1234:8080` -- accessing `localhost:1234` allows accessing `8080` of `my-instance`.
- `kraft cloud tunnel my-instance 1234:8080/udp`, same as above, only for UDP.

### PR DESCRIPTION
**Pull request by craciunoiuc**
### Prerequisite checklist

  - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [ ] Tested your changes against relevant architectures and platforms;
  - [ ] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

To test: `kraft cloud tunnel lport:instance:rport`

GitHub-Closes: #1495

### CODE REVIEW

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 1]
```diff
@@ -89,6 +91,8 @@ func (c *connection) handle(ctx context.Context) {
 	}
 	defer rc.Close()
 
+	log.G(ctx).Info("Accepted client connection ", c.conn.RemoteAddr(), " to ", rc.LocalAddr(), "->", rc.RemoteAddr())
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]:
```diff
- log.G(ctx).Info("Accept
+ log.G(ctx).Info("accept
```
Lower case first character for log messages

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 2]
```diff
@@ -99,7 +103,13 @@ func (c *connection) handle(ctx context.Context) {
 		_ = c.conn.SetDeadline(immediateNetCancel)
 	}()
 
-	const bufSize = 32 * 1024 // same as io.Copy
+	if len(auth) > 0 {
+		_, err = rc.Write(auth)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("Failed to write auth to remote host")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]: Same here

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 3]
```diff
@@ -108,36 +118,25 @@ func (c *connection) handle(ctx context.Context) {
...
+		_, err = io.Copy(rc, c.conn)
+		if err != nil {
+			if isNetClosedError(err) {
 				return
 			}
-			if _, err := rc.Write(writeBuf[:n]); err != nil {
-				log.G(ctx).WithError(err).Error("Failed to write to remote host")
-				return
+			if !isNetTimeoutError(err) {
+				log.G(ctx).WithError(err).Error("Failed to copy data from client to remote host")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]: And here

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 4]
```diff
@@ -108,36 +118,25 @@ func (c *connection) handle(ctx context.Context) {
...
-			}
-			break
-		}
-		if _, err := c.conn.Write(readBuf[:n]); err != nil {
-			log.G(ctx).WithError(err).Error("Failed to write to client")
-			break
+	_, err = io.Copy(c.conn, rc)
+	if err != nil {
+		if !isNetTimeoutError(err) {
+			log.G(ctx).WithError(err).Error("Failed to copy data from remote host to client")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]: One more time

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 5]
```diff
@@ -46,7 +49,29 @@ func (r *Relay) Up(ctx context.Context) error {
...
+		return err
+	}
+	defer rc.Close()
+	go func() { <-ctx.Done(); rc.Close() }()
+
+	ready <- struct{}{}
+	close(ready)
+
+	// Heartbeat every minute to keep the connection alive
+	heartbeat := []byte("\xf0\x9f\x91\x8b\xf0\x9f\x90\x92\x00")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]: Define `heartbeat` as `const` and expose as `const HeartBeat = ...`

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 6]
```diff
@@ -109,6 +134,36 @@ func (c *connection) handle(ctx context.Context, auth []byte) {
 			log.G(ctx).WithError(err).Error("Failed to write auth to remote host")
 			return
 		}
+
+		var status []byte
+		statusRaw := bytes.NewBuffer(status)
+		n, err := io.CopyN(statusRaw, rc, 2)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("Failed to read auth status from remote host")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]: Lowercase first char

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 7]
```diff
@@ -109,6 +134,36 @@ func (c *connection) handle(ctx context.Context, auth []byte) {
...
+
+		if n != 2 {
+			log.G(ctx).Error("Invalid auth status from remote host")
+			return
+		}
+
+		var statusParsed int16
+		err = binary.Read(statusRaw, binary.LittleEndian, &statusParsed)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("Failed to parse auth status from remote host")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]: Lowercase first char

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 8]
```diff
@@ -109,6 +134,36 @@ func (c *connection) handle(ctx context.Context, auth []byte) {
...
+			return
+		}
+
+		if statusParsed == 0 {
+			log.G(ctx).Error("No more available connections to remote host. Try again later")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]: Lowercase first char

File: internal/cli/kraft/cloud/tunnel/relay.go [hunk 9]
```diff
@@ -109,6 +134,36 @@ func (c *connection) handle(ctx context.Context, auth []byte) {
...
+			return
+		} else if statusParsed < 0 {
+			log.G(ctx).Errorf("Internal tunnel error (C=%d), to view logs run:", statusParsed)
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/relay.go]: Lowercase first char

File: internal/cli/kraft/cloud/tunnel/tunnel.go [hunk 10]
```diff
@@ -11,31 +11,70 @@ import (
...
+	"kraftkit.sh/log"
 )
 
 type TunnelOptions struct {
-	token string
-	metro string
+	Connections      uint     `local:"true" long:"connections" short:"c" usage:"Number of connections to forward maximum at the same time" default:"27"`
+	ProxyPorts       []string `local:"true" long:"proxy-ports" short:"p" usage:"Ports to use for the proxies. Default start port is 4444"`
+	ProxyControlPort uint     `local:"true" long:"proxy-control-port" short:"P" usage:"Port to use for the proxy control" default:"4443"`
+
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/tunnel.go]: Delete the newline between `ProxyControlPort` and `Token`

File: internal/cli/kraft/cloud/tunnel/tunnel.go [hunk 11]
```diff
@@ -49,99 +88,317 @@ func NewCmd() *cobra.Command {
...
 
 func (opts *TunnelOptions) Run(ctx context.Context, args []string) error {
-	sgID := args[0]
+	var err error
 
-	lport, rport, err := parsePorts(args[1])
-	if err != nil {
-		return err
+	if len(opts.ProxyPorts) > 1 && len(opts.ProxyPorts) != len(args) {
+		return fmt.Errorf("custom number of proxy ports must match the number of ports to forward")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/tunnel.go]:
```diff
- return fmt.Errorf("custom number of proxy ports must match the number of ports to forward")
+ return fmt.Errorf("supplied number of proxy ports must match the number of ports to forward")
```

File: internal/cli/kraft/cloud/tunnel/tunnel.go [hunk 12]
```diff
@@ -49,99 +88,317 @@ func NewCmd() *cobra.Command {
...
+	if len(types) == 2 {
+		ctype = types[1]
+	} else {
+		ctype = "tcp"
 	}
 
-	if len(sg.Domains) == 0 {
-		return "", fmt.Errorf("service group '%s' has no public domain", sgID)
+	if strings.ToLower(ctype) != "tcp" {
+		log.G(ctx).Warn("Only TCP connections are supported at the moment")
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/tunnel.go]: Lowercase first char

File: internal/cli/kraft/cloud/tunnel/tunnel.go [hunk 13]
```diff
@@ -49,99 +88,317 @@ func NewCmd() *cobra.Command {
...
+		Port:            int(proxyControlPort),
+		DestinationPort: ptr(int(proxyControlPort)),
+		Handlers: []kcservices.Handler{
+			kcservices.HandlerTLS,
+		},
+	})
 
-	rport64, err := strconv.ParseUint(ports[1], 10, 16)
+	crinstResp, err := cli.Create(ctx, kcinstances.CreateRequest{
+		Image:    "kraftcloud/tunlr:latest",
```
Comment (nderjung) [inline internal/cli/kraft/cloud/tunnel/tunnel.go]: Let's define this image as a `const` and expose publicly as `TunnelServiceImage`.

Comment (nderjung) [reply internal/cli/kraft/cloud/tunnel/tunnel.go]: Additionally, this should be configurable as a CLI option, `--tunnel-service-image`.

**Review by skuenzer — APPROVED** [review]
Reviewed-by: Simon Kuenzer <simon@unikraft.io>

**Pull request merged**